api/external_auth.proto

```diff
@@ -0,0 +1,151 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2;
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with OK or NOT OK.
+service Authorization {
+  rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+  // The name of the tenant for multi-tenant support.
+  string tenant_name = 1;
+
+  // The request attributes.
+  AttributeContext attributes = 2;
+}
+
+message CheckResponse {
+  // Status OK allows the request. Any other response indicates the request should be denied.
+  google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// The fields follow the Istio attribute vocabulary.
+// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+// The field type becomes the attribute type, and the field path becomes the attribute name.
+// For example, the attribute `source.ip` maps to field `AttributeContext.source.ip`.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The IP address of the peer.
+    string ip = 1;
+
+    // The network port of the peer.
+    int64 port = 2;
```

Comment:
We're definitely getting into API philosophical differences between Envoy and Istio here. In Envoy, we would make this an unsigned type to start with (it's the Envoy API convention, regardless of the signed vs. unsigned debate), and probably a wrapped type, as not all peers will be on ports.

Taking this a bit further, we already have our own Address type, that capture all the different ways a peer can be connected, see https://github.com/envoyproxy/data-plane-api/blob/master/api/address.proto. Should we just use that? Same as with logging.

Reply:
I'm with @htuch on reusing the `Address` message.  It's easy to understand and handles non-IP peers (e.g. Unix Domain Sockets).

Reply:
+1 to using Address, as it will correctly deal with UDS as well as UDP when we implement that. It's true that some fields don't really apply such as resolver, but we just won't set them.